README.md
AstDemo
=======

XMPP + VoIP (Openfire + Asterisk) demo integration.

Requisites
**********
* Openfire running
* Asterisk 11 or later running with XMPP support
* PostgreSQL running
* ServerInfo Plugin Support (http://www.mundoopensource.com.br/serverinfo-plugin-openfire/)
* python pyst (http://sourceforge.net/projects/pyst/)
* python psycopg2 (http://initd.org/psycopg/)
* python xmpppy (http://xmpppy.sourceforge.net/)

Installation
************
* Create a subdirectory in your agi-bin asterisk dir named astdemo
* Put all astdemo files there
* Create you astdemo database (you can use db/pgsql/astdemo.sql)
* Copy config.ini.sample to config.ini and update inifile with your data
* Configure your xmpp Asterisk resource (asterisk_config/xmpp.conf has a sample of configuration)
* Edit your extensions.conf file, adding [from_xmpp_demo] context, [call] context and adding variables in [globals] context. If you want support to features like callerid and dnd, adapt your extensions context too. See extensions.conf file in asterisk_config directory.

Available Commands
******************
* callmsg <number> - make a call to <number> and play tt-monkeys when answered.
* callerid <extension> [jid] - enable or disable the callerid xmpp message on received calls to extension.
* info <voip|xmpp> - show VoIP or XMPP server informations.
* help - show this help.
* iaxpeers - show IAX peers.
* monitor iax <peer> - monitor iax <peer> state.
* monitor sip <extension> - monitor <extension> state.
* monitor queue <queue> - monitor queue <queue> and alarm when it has no members logged.
* monitoring - show monitoring items.
* sippeer <extension> - show SIP informations of <extension>.
* sippeers - show SIP peers.
* unmonitor iax <peer> - unmonitor iax <peer> state.
* unmonitor sip <extension> - unmonitor <extension> state.
* unmonitor queue <queue> - unmonitor queue <queue>.
* version - show astdemo version.

If you have Asterisk Realtime enabled you'll have access to more commands:
* call <sip extension> - call to <sip extension> and transfer to your extension when answered.
* dnd <on|off|status> - enable/disable or show Do Not Disturb status.
* queue <queue> - show <queue> members and statistics.
* queuemsg <queue> <message> - send <message> to all members of <queue>.
* enter <queue> - join <queue>.
* leave <queue> - leave <queue>.
* whoami - show informations about myself.

Asterisk Realtime Support
*************************
First question: do you have Asterisk Realtime enabled? Read more about Asterisk Realtime Architecture in https://wiki.asterisk.org/wiki/display/AST/Realtime+Database+Configuration. After your Asterisk have Realtime support, you can continue.

To activate Asterisk Realtime support in AstDemo, first you need to have a jid field in your sip database table (jid, character varying, length 150). 
If you have it, you just need to enable asterisk realtime support in config.ini file. 

If you activate Asterisk Realtime support in AstDemo you will release more AstDemo features.

After Asterisk Realtime support activation you will see new commands in help and some commands will change, like callerid, cause you don't need to indicate your extension and your jid anymore, it'll just toggle xmpp message. 

ATTENTION: you need to configure your sip record with your jid, or new resources will not work.

Monitoring
**********
To activate monitoring copy the cronjob file (cron/astdemo_monitor) to /etc/cron.d. 
To monitor IAX or SIP devices you need to qualify peer in its configuration (qualify=yes)


CHANGELOG
* Version 0.1.6.2 (2014-01-03) Marcelo H. Terres <mhterres@mundoopensource.com.br>
** astdemo.sql file changed (owner correction)
** README file changed (xmpppy requirement added)
-----------------------------------

*
* Version 0.1.6.1 (2014-12-30) Marcelo H. Terres <mhterres@mundoopensource.com.br>
** Command call changed to callmsg
** New command call (call SIP extension and transfer to your extension after answered)
** New directory (audios) ans file transfer.sln (googletts generated)
-----------------------------------

* Version 0.1.6 (2014-12-30) Marcelo H. Terres <mhterres@mundoopensource.com.br>
** README.md changed
** Code sanitization
** Command monitor sip added
** Command monitor queue added
** Command unmonitor sip added
** Command unmonitor queue added
** Command monitoring added
** Database updated to support monitor features
** Added monitor cronjob
** Added monitor script
-----------------------------------

* Version 0.1.5.2 (2014-12-29) Marcelo H. Terres <mhterres@mundoopensource.com.br>
** Change command queueon to enter
** Change command queueoff to leave
-----------------------------------


* Version 0.1.5.1 (2014-12-28) Marcelo H. Terres <mhterres@mundoopensource.com.br>
** README.md improvement
---------------------------------


* Version 0.1.5 (2014-12-28) Marcelo H. Terres <mhterres@mundoopensource.com.br>
** PostgresSQL schema added
** Added support for Asterisk Realtime
** Improved documentation
** Improved source and directory organization
** Command whoami added (asterisk realtime required)
** Command dnd added (asterisk realtime required)
** Command queuemsg added (asterisk realtime required)
** Command queue added (asterisk realtime required)
** Command queueon added (asterisk realtime required)
** Command queueoff added (asterisk realtime required)
---------------------------------


* Version 0.1.4.1 (2014-12-27) Marcelo H. Terres <mhterres@mundoopensource.com.br>
** TODO added
---------------------------------

* Version 0.1.4 (2014-12-26) Marcelo H. Terres <mhterres@mundoopensource.com.br>
** New feature added: callerid
---------------------------------

* Version 0.1.3.2 (2014-12-26) Marcelo H. Terres <mhterres@mundoopensource.com.br>
** CALLS.txt added
---------------------------------

* Version 0.1.3.1 (2014-12-26) Marcelo H. Terres <mhterres@mundoopensource.com.br>
** Removed config.ini
** README.me updated
---------------------------------

* Version 0.1.3 (2014-12-26) Marcelo H. Terres <mhterres@mundoopensource.com.br>
** First released version
---------------------------------


xmpp_defs.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

# Defs
#
# Marcelo H. Terres <mhterres@gmail.com>
# 2014-06-02
#
# Versao 0.2.1.1

import socket
import time
import re
import psycopg2
import psycopg2.extras
import ConfigParser

from xmpp_classes import connectAMI
from xmpp_classes import activeLogs
from xmpp_classes import ManagerDict
from xmpp_classes import ManagerDictEvents

def sipPeer(sip):

	activeLogs('/tmp','xmpp_defs','err')

	myConnect=connectAMI()
	s = myConnect.socket

	s.send('Action: SIPshowpeer\nPeer: ' + sip + '\n\n')
	time.sleep (0.2)

	data = s.recv(65536)

	mgrdict=ManagerDict(data)

	returnmsg = "Extension %s informations: " % sip
	returnmsg += "\r"

	if data.find("Peer " + sip + " not found") > 0:

		returnmsg += "Ramal " + sip + " não registrado.\n"
	else:		
		returnmsg += "CallerID: " + mgrdict.getitem('Callerid') 
		returnmsg += "IP: " + mgrdict.getitem('Address-IP')
		returnmsg += "Codecs: " + mgrdict.getitem('Codecs')
		returnmsg += "Status: " + mgrdict.getitem('Status')
		returnmsg += "User Agent: " + mgrdict.getitem('SIP-Useragent')

	s.close()

	return returnmsg

def infoVoIP():

	activeLogs('/tmp','xmpp_defs','err')

	myConnect=connectAMI()
	s = myConnect.socket

	s.send('Action: CoreSettings\n\n')
	time.sleep (0.2)

	data = s.recv(65536)

	mgrdict=ManagerDict(data)

	returnmsg = "VoIP Informations"
	returnmsg += "\r"
	returnmsg += "Asterisk Version %s " % mgrdict.getitem('AsteriskVersion') 

	s.send('Action: CoreStatus\n\n')
	time.sleep (0.2)

	data = s.recv(65536)

	mgrdict=ManagerDict(data)

	returnmsg += "Started: " + mgrdict.getitem('CoreStartupDate').replace('\r','') + ' - ' + mgrdict.getitem('CoreStartupTime')
	returnmsg += "Last reload: " + mgrdict.getitem('CoreReloadDate').replace('\r','')  + ' - ' + mgrdict.getitem('CoreReloadTime')

	s.close()
	
	return returnmsg

def iaxPeers():

	myConnect=connectAMI()
	s = myConnect.socket

	s.send('Action: IAXpeers\n\n')
	time.sleep (0.2)

	data = s.recv(65535)

	mgrdictev=ManagerDictEvents(data,'PeerEntry')

	if not mgrdictev.isvalid:

		returnmsg = "IAX peers not found."
		return returnmsg
	
	else:

		returnmsg = "IAX peers Informations - %s peers found.\r" % str(mgrdictev.items) 

		for event in mgrdictev.events:


			returnmsg+="Peer: " + event['ObjectName'] 
			returnmsg+="Type: " + event['ChanObjectType'] 
			returnmsg+="IP: " + event['IPaddress'] 
			returnmsg+="Port: " + event['IPport'] 
			returnmsg+="Trunk: " + event['Trunk'] 
			returnmsg+="Status: " + event['Status'] 
			returnmsg += '\r'

	s.close()
	
	return returnmsg

def sipPeers():

	myConnect=connectAMI()
	s = myConnect.socket

	s.send('Action: SIPpeers\n\n')
	time.sleep (0.2)

	data = s.recv(65535)

	mgrdictev=ManagerDictEvents(data,'PeerEntry')

	if not mgrdictev.isvalid:

		returnmsg = "SIP peers not found."
		return returnmsg
	
	else:

		returnmsg = "SIP peers Informations - %s peers found.\r " % str(mgrdictev.items) 

		for event in mgrdictev.events:


			returnmsg+="Peer: " + event['ObjectName'] 
			returnmsg+="Type: " + event['ChanObjectType'] 
			returnmsg+="IP: " + event['IPaddress'] 
			returnmsg+="Port: " + event['IPport'] 
			returnmsg+="Video support: " + event['VideoSupport'] 
			returnmsg+="Realtime: " + event['RealtimeDevice'] 
			returnmsg+="Status: " + event['Status'] 
			returnmsg += '\r'

	s.close()
	
	return returnmsg



astdemo_db.py
#!/usr/bin/python

import re
import psycopg2
import psycopg2.extras
import sys
from astdemo_classes import Config
from astdemo_classes import activeLogs

class DB:

	def __init__(self):

		activeLogs('/tmp','astdemo_db','err')

		cfg = Config()

		dsn = 'dbname=%s host=%s user=%s password=%s' % (cfg.db_dbname, cfg.db_host, cfg.db_user, cfg.db_pwd)

		conn = psycopg2.connect(dsn)
		curs = conn.cursor(cursor_factory=psycopg2.extras.DictCursor)
		self.cursor = curs
		self.config = cfg

	def logMessage(self,msg_from,msg_to,msg_msg):

		sql = "BEGIN TRANSACTION; INSERT INTO logs VALUES(DEFAULT,DEFAULT,%s,%s,%s); COMMIT;"
		self.cursor.execute(sql, (msg_from,msg_to,msg_msg,))

	def monitorOperation(self,jid,operation,operation_type,operation_target):

		if operation=="monitor":

			if not self.monitorOperationExists(jid,operation,operation_type,operation_target):

				sql = "BEGIN TRANSACTION; INSERT INTO monitors VALUES(DEFAULT,DEFAULT,%s,%s,%s,%s); COMMIT;"
				self.cursor.execute(sql, (jid,operation,operation_type,operation_target))

				if self.config.monitoring_log_operation == "1":

					sql = "BEGIN TRANSACTION; INSERT INTO monitors_logs VALUES(DEFAULT,DEFAULT,%s,%s,%s,%s); COMMIT;"
					self.cursor.execute(sql, (jid,operation,operation_type,operation_target))

				return True

			else:

				return False

		else:

				if self.monitorOperationExists(jid,operation,operation_type,operation_target):

					sql = "BEGIN TRANSACTION; DELETE FROM monitors WHERE jid=%s AND operation='monitor' AND type=%s AND target=%s;COMMIT;"
					self.cursor.execute(sql, (jid,operation_type,operation_target))

					if self.config.monitoring_log_operation == "1":

						sql = "BEGIN TRANSACTION;INSERT INTO monitors_logs VALUES(DEFAULT,DEFAULT,%s,%s,%s,%s);COMMIT;"
						self.cursor.execute(sql, (jid,operation,operation_type,operation_target))

					return True

				else:

					return False

	def monitorOperationExists(self,jid,operation,operation_type,operation_target):

		sql = "SELECT * from monitors WHERE jid=%s AND operation='monitor' AND type=%s AND target=%s;"
		self.cursor.execute(sql, (jid,operation_type,operation_target))

		if not self.cursor.rowcount:

			return False
			
		return True

	def monitorUpdateState(self,jid,operation_type,operation_target,operation_state):

		if self.monitorOperationExists(jid,'monitor',operation_type,operation_target):

			sql = "BEGIN TRANSACTION;UPDATE monitors SET laststate=%s WHERE jid=%s AND operation='monitor' AND type=%s AND target=%s;COMMIT;"
			self.cursor.execute(sql, (operation_state,jid,operation_type,operation_target))

			if self.config.monitoring_log_state == "1":

				sql = "BEGIN TRANSACTION;INSERT INTO monitors_logs VALUES(DEFAULT,DEFAULT,'',%s,%s,%s,%s);COMMIT;"
				self.cursor.execute(sql, ('monitor',operation_type,operation_target,operation_state))

	def monitoringItems(self,jid):

		sql = "SELECT date,operation,type,target,laststate FROM monitors WHERE jid='%s';" % jid
		self.cursor.execute(sql)

		returnmsg=""

		if self.cursor.rowcount:

			for row in self.cursor:

					returnmsg+="%s - %s %s %s - last status: %s\r" % (row['date'],row['operation'].strip(),row['type'].strip(),row['target'].strip(),row['laststate'])

		return returnmsg

	def returnAllItems(self):

		sql = "SELECT * FROM monitors;"
		self.cursor.execute(sql)

		return (self.cursor)


xmpp_classes.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

# Classes
#
# Marcelo H. Terres <mhterres@gmail.com>
# 2014-06-02
#
# Versao 0.2.1.1

import os
import sys
import xmpp
import time
import socket
import ConfigParser

class activeLogs:

	def __init__(self,direct,filename,typelog):

		log = open(direct + '/' + filename + '.log' , 'a')

		if typelog == 'err':

			sys.stderr = log

		elif typelog == 'out':

			sys.stdout = log

		elif typelog == 'all':

			sys.stderr = log
			sys.stdout = log

class connectAMI:

	def __init__(self):

		myConfigs = Config()

		s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
		s.connect((myConfigs.voip_manager_srv,5038))

		s.send('Action: Login\nUsername: '+myConfigs.voip_manager_usr+'\nSecret: '+myConfigs.voip_manager_pwd+'\nEvents: off\n\n')
		time.sleep (0.2)
		data = s.recv(2048)

                self.socket = s

class ManagerDict:

	def __init__(self,text):

		activeLogs('/tmp','xmpp_classes','err')

		# split the text
		linhas = text.split('\n')

		d={}

		for item in linhas:

			dado=item.split(":",1)

			try:

				title=dado[0]
			except:

				title=""

			try:

				value=dado[1]
			except:

				value=""


			d.update({title:value})

		self.dicti=d

	def getitem(self,index):

		i=self.dicti[index]
		i=i.replace(" ","\ ")
		return i

class Config:

	def __init__(self):

		self.description = 'AstDemo configurations'
		activeLogs('/tmp','xmpp_classes','err')

		# Read PGSql configuration and connect to there
		configuration = ConfigParser.RawConfigParser()
		configuration.read('%s/config.ini' % os.path.dirname(sys.argv[0]))

		# script
		self.version=configuration.get('general','version')

		# xmpp
		self.jid_auth=configuration.get('general','jid_auth')
		self.jid_pwd=configuration.get('general','jid_pwd')

		# voip
		self.voip_manager_srv=configuration.get('general','voip_manager_srv')
		self.voip_manager_usr=configuration.get('general','voip_manager_usr')
		self.voip_manager_pwd=configuration.get('general','voip_manager_pwd')

		# db
		self.db_host=configuration.get('general','db_host')
		self.db_dbname=configuration.get('general','db_dbname')
		self.db_user=configuration.get('general','db_user')
		self.db_pwd=configuration.get('general','db_pwd')

		# serverinfo
		self.serverinfo_host=configuration.get('general','serverinfo_host')
		self.serverinfo_port=configuration.get('general','serverinfo_port')
	
class XMPPp:

	def __init__(self):


		activeLogs('/tmp/','xmpp_classes','err')

		myConfigs=Config()

		jid = xmpp.protocol.JID(myConfigs.jid_auth)

		self.domain=jid.getDomain()
		self.node=jid.getNode()

		cl=xmpp.Client(self.domain,debug=[])
		con=cl.connect()

		if not con:

			self.connected=False
		else:

			self.connected=True

			auth=cl.auth(self.node,myConfigs.jid_pwd)

			if not auth:

				self.authenticated=False

			else:

				self.authenticated=True


class ManagerDictEvents:

	def __init__(self,text,myevent):

		self.description = "Generate dictionary of a specific event."

		activeLogs('/tmp','xmpp_classes','err')

		# split the text
		linhas = text.split('\n')

		i=0

		dicio={}
		events=[]

		fulldicio={}
		dicioitem=0

		if text.find('Event: ' + myevent) == -1:

			self.isvalid=False
		else:

			self.isvalid=True

			self.events=[]
			self.items=0
			self.fulldicti={}
			self.dictiitems=0

			for myitem in linhas:

				dado=myitem.split(":",1)

				try:

					title=dado[0]
				except:

					title=""

				try:

					value=dado[1]
				except:

					value=""

				if (title == "Event" and value.find(myevent) > 0 and len(dicio) == 0):

					i+=1
					events.append({})
					dicioitem+=1
					fulldicio.update({title:value})

					dicio.update({title:value})
					events[i-1].update({title:value})
			
				elif (title == "Event" and len(dicio)>0):

					dicioitem+=1
					fulldicio.update({title:value})

					if len(dicio) > 0:

						events[i-1].update({title:value})

						dicio.clear()
						dicio={}

					if value.find(myevent) > 0:

						i+=1
						events.append({})
						dicioitem+=1
						fulldicio.update({title:value})
						
						events[i-1].update({title:value})
						dicio.update({title:value})

				elif (title != "" and len(dicio)>0):

					dicioitem+=1
					fulldicio.update({title:value})

					events[i-1].update({title:value})
					dicio.update({title:value})

			self.items=i
			self.events=events
			self.dicti=fulldicio
			self.dictiitems=dicioitem
			self.text=text




astdemo_monitor.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

# AstDemo_monitor
# Monitoring script
#
# Marcelo H. Terres <mhterres@gmail.com>
# 2014-12-30
#

import xmpp

from asterisk import agi

from astdemo_defs import iaxPeerStatus
from astdemo_defs import sipPeerStatus
from astdemo_defs import getQueueMembersNumber

from astdemo_db	import DB
from astdemo_classes import XMPPp
from astdemo_classes import Config
from astdemo_classes import activeLogs

activeLogs('/tmp','astdemo_monitor','all')

db=DB()
cfg=Config()

xmppdomain = cfg.xmppdomain
xmppresource = cfg.xmppresource

myXMPP=XMPPp()

db_cursor=db.returnAllItems()

totalRows=db_cursor.rowcount

print "-----------"

if totalRows:

	print "%i items monitored." % totalRows

	x = 1

	rows=db_cursor.fetchall()

	for row in rows:
	
		jid=row['jid'].strip()
		op_type=row['type'].strip()
		op_target=row['target'].strip()
		laststate=row['laststate']

		print "Processing row %i - %s %s" % (x,op_type,op_target)

		x = x + 1

		if op_type=="sip":

			state=sipPeerStatus(op_target).strip()
			message="SIP %s is now %s." % (op_target,state)

		elif op_type=="iax":

			state=iaxPeerStatus(op_target).strip()
			message="IAX %s is now %s." % (op_target,state)

		elif op_type=="queue":

			members=getQueueMembersNumber(op_target)

			if members==0:

				state="no members"
			else:

				state="%i member(s)" % members

			message="Queue %s now has %s." % (op_target,state)

		print state
		if laststate != state:

				db.monitorUpdateState(jid,op_type,op_target,state)
				myXMPP.client.send(xmpp.protocol.Message(jid,message))




astdemo_classes.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

# AstDemo_classes
#
# Marcelo H. Terres <mhterres@gmail.com>
# 2014-06-02
#
# Updated 2014/12/30

import os
import sys
import xmpp
import time
import socket
import ConfigParser

class activeLogs:

	def __init__(self,direct,filename,typelog):

		log = open(direct + '/' + filename + '.log' , 'a')

		if typelog == 'err':

			sys.stderr = log

		elif typelog == 'out':

			sys.stdout = log

		elif typelog == 'all':

			sys.stderr = log
			sys.stdout = log

class connectAMI:

	def __init__(self):

		myConfigs = Config()

		s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
		s.connect((myConfigs.voip_manager_srv,5038))

		s.send('Action: Login\nUsername: '+myConfigs.voip_manager_usr+'\nSecret: '+myConfigs.voip_manager_pwd+'\nEvents: off\n\n')
		time.sleep (0.1)
		data = s.recv(2048)

		self.socket = s

class ManagerDict:

	def __init__(self,text):

		activeLogs('/tmp','astdemo_classes','err')

		# split the text
		lines = text.split('\n')

		d={}

		for item in lines:

			data=item.split(":",1)

			try:

				title=data[0]
			except:

				title=""

			try:

				value=data[1]
			except:

				value=""


			d.update({title:value})

		self.dicti=d

	def getitem(self,index):

		i=self.dicti[index]
		i=i.replace(" ","\ ")
		return i

class Config:

	def __init__(self):

		self.description = 'AstDemo configurations'
		activeLogs('/tmp','astdemo_classes','err')

		# Read PGSql configuration and connect to there
		configuration = ConfigParser.RawConfigParser()
		configuration.read('%s/config.ini' % os.path.dirname(sys.argv[0]))

		# script
		self.path = os.path.dirname(sys.argv[0])
		self.version=configuration.get('general','version')

		# xmpp
		self.jid_auth=configuration.get('general','jid_auth')
		self.jid_pwd=configuration.get('general','jid_pwd')
		self.xmppdomain=configuration.get('general','xmppdomain')
		self.xmppresource=configuration.get('general','xmppresource')

		# voip
		self.voip_manager_srv=configuration.get('general','voip_manager_srv')
		self.voip_manager_usr=configuration.get('general','voip_manager_usr')
		self.voip_manager_pwd=configuration.get('general','voip_manager_pwd')

		# db
		self.db_host=configuration.get('general','db_host')
		self.db_dbname=configuration.get('general','db_dbname')
		self.db_user=configuration.get('general','db_user')
		self.db_pwd=configuration.get('general','db_pwd')

		# serverinfo
		self.serverinfo_host=configuration.get('general','serverinfo_host')
		self.serverinfo_port=configuration.get('general','serverinfo_port')

		# asterisk
		self.asterisk_realtime=configuration.get('general','asterisk_realtime')
		self.asterisk_db_host=configuration.get('general','asterisk_db_host')
		self.asterisk_db_name=configuration.get('general','asterisk_db_name')
		self.asterisk_db_user=configuration.get('general','asterisk_db_user')
		self.asterisk_db_pwd=configuration.get('general','asterisk_db_pwd')
		self.asterisk_sip_table=configuration.get('general','asterisk_sip_table')
		self.asterisk_jid_field=configuration.get('general','asterisk_jid_field')

		# monitoring
		self.monitoring_log_operation=configuration.get('general','monitoring_log_operation')
		self.monitoring_log_state=configuration.get('general','monitoring_log_state')

		# call cmd
		self.call_callerid = configuration.get('general','call_callerid')
		self.call_context = configuration.get('general','call_context')

class XMPPp:

	def __init__(self):


		activeLogs('/tmp/','astdemo_classes','err')

		myConfigs=Config()

		jid = xmpp.protocol.JID(myConfigs.jid_auth)

		self.domain=jid.getDomain()
		self.node=jid.getNode()

		cl=xmpp.Client(self.domain,debug=[])
		con=cl.connect()

		if not con:

			self.connected=False
		else:

			self.connected=True

			auth=cl.auth(self.node,myConfigs.jid_pwd)

			if not auth:

				self.authenticated=False

			else:

				self.authenticated=True

		self.client = cl


class ManagerDictEvents:

	def __init__(self,text,myevent):

		self.description = "Generate dictionary of a specific event."

		activeLogs('/tmp','astdemo_classes','err')

		# split the text
		lines = text.split('\n')

		i=0

		dicio={}
		events=[]

		fulldicio={}
		dicioitem=0

		if text.find('Event: ' + myevent) == -1:

			self.isvalid=False
		else:

			self.isvalid=True

			self.events=[]
			self.items=0
			self.fulldicti={}
			self.dictiitems=0

			for myitem in lines:

				data=myitem.split(":",1)

				try:

					title=data[0]
				except:

					title=""

				try:

					value=data[1]
				except:

					value=""

				if (title == "Event" and value.find(myevent) > 0 and len(dicio) == 0):

					i+=1
					events.append({})
					dicioitem+=1
					fulldicio.update({title:value})

					dicio.update({title:value})
					events[i-1].update({title:value})
			
				elif (title == "Event" and len(dicio)>0):

					dicioitem+=1
					fulldicio.update({title:value})

					if len(dicio) > 0:

						events[i-1].update({title:value})

						dicio.clear()
						dicio={}

					if value.find(myevent) > 0:

						i+=1
						events.append({})
						dicioitem+=1
						fulldicio.update({title:value})
						
						events[i-1].update({title:value})
						dicio.update({title:value})

				elif (title != "" and len(dicio)>0):

					dicioitem+=1
					fulldicio.update({title:value})

					events[i-1].update({title:value})
					dicio.update({title:value})

			self.items=i
			self.events=events
			self.dicti=fulldicio
			self.dictiitems=dicioitem
			self.text=text

class Queue:

		def __init__(self,queue,agi):

				self.description = 'Queue status'

				activeLogs('/tmp','astdemo_classes','err')

				myConnect=connectAMI()
				s = myConnect.socket

				astdb=astDB(agi)
				self.astdb=astdb

				s.send('Action: QueueSummary\nQueue: ' + queue + '\n\n')
				time.sleep (0.1)

				data = s.recv(65536)

				if data.find('Queue: ' + queue) == -1:

						self.isavailable=False
				else:

						self.isavailable=True

						mgrdict=ManagerDict(data)

						self.name=queue
						self.loggedmembers=mgrdict.getitem('LoggedIn')
						self.availablemembers=mgrdict.getitem('Available')
						self.members=[]

				s.close

		def membersNumber(self):

				members=0

				myConnect=connectAMI()
				s = myConnect.socket

				s.send('Action: QueueStatus\nQueue: ' + self.name + '\n\n')
				time.sleep (0.1)

				data = s.recv(65536)
				mgrdict=ManagerDict(data)

				lines = data.split('\n')

				for item in lines:

						data=item.split(":",1)

						try:

								title=data[0]
						except:

								title=""

						if title == "Name":

								members = members + 1

				s.close

				return (members)


		def getmembers(self):

				myConnect=connectAMI()
				s = myConnect.socket

				s.send('Action: QueueSummary\nQueue: %s\n\n' % self.name)
				time.sleep (0.1)
				data = s.recv(65536)

				mgrdict=ManagerDict(data)

				self.loggedmembers=mgrdict.getitem('LoggedIn')
				self.availablemembers=mgrdict.getitem('Available')

				s.send('Action: QueueStatus\nQueue: ' + self.name + '\n\n')
				time.sleep (0.1)

				data2 = s.recv(65536)
				mgrdict=ManagerDict(data2)

				self.strategy = mgrdict.getitem('Strategy')
				self.calls = mgrdict.getitem('Calls')
				self.completed = mgrdict.getitem('Completed')
				self.abandoned = mgrdict.getitem('Abandoned')
				self.holdtime = mgrdict.getitem('Holdtime')
				self.talktime = mgrdict.getitem('TalkTime')

				lines = data2.split('\n')

				aMembers=[]

				for item in lines:

						data=item.split(":",1)

						try:

								title=data[0]
						except:

								title=""

						try:

								value=data[1]
						except:

								value=""

						if title == "Name":

								aMembers.append(value)

				self.members=[]

				self.members=aMembers

				s.close

				return (aMembers)

		def ismember(self,member):

				returnValue=False

				for item in self.members:

						if 'SIP/' + member in item:

								returnValue=True

				return returnValue

		def addmember(self,member,agi):

				myConnect=connectAMI()
				s = myConnect.socket

				if self.ismember(member):

						returnmsg = "Extension %s is already a queue %s member." % (member,self.name)
				else:

						s.send('Action: QueueAdd\nQueue: ' + self.name + '\nInterface: SIP/' + member + '\n\n')
						time.sleep(.2)

						self.getmembers()

						if self.ismember(member):

							self.astdb.put('queue',member,self.name)
							self.astdb.put('queuelogin',member,1)

							returnmsg = "Extension %s now is a member of queue %s." % (member,self.name)
						else:
	
							returnmsg = "Error: extension %s can't join queue %s." % (member,self.name)
							s.close()

				return (returnmsg)

		def removemember(self,member,agi):

				myConnect=connectAMI()
				s = myConnect.socket

				if not self.ismember(member):

					returnmsg = "Extension %s is not a queue %s member." % (member,self.name)

				else:

						s.send('Action: QueueRemove\nQueue: ' + self.name + '\nInterface: SIP/' + member + '\n\n')
						time.sleep(.2)

						self.getmembers()

						if not self.ismember(member):

								self.astdb.rem('queue',member)
								self.astdb.rem('queuelogin',member)

								returnmsg = "Extension %s leaves queue %s." % (member,self.name)
						else:

								returnmsg = "Error: extension %s can't leave queue %s." % (member,self.name)


				s.close()

				return (returnmsg)

class astDB():

	def __init__(self,agi):

		self.description = 'Asterisk Database'

		activeLogs('/tmp','astdemo_classes','err')

		self.agi=agi

	def get(self,key,extension):

		return self.agi.database_get("exten_%s" % extension, key)

	def put(self,key,extension,value):

		self.agi.database_put("exten_%s" % extension, key, value)

	def rem(self,key,extension):

		self.agi.database_del("exten_%s" % extension, key);

